cpp/src/arrow/acero/scalar_aggregate_node.cc

```diff
@@ -294,6 +294,14 @@ Status ScalarAggregateNode::OutputResult(bool is_last) {
   // First, insert segment keys
   PlaceFields(batch, /*base=*/0, segmenter_values_);
 
+  // Move away the states and recreate them eagerly, to make sure that any error
+  // below does not leave us with empty states.
+  auto states = std::move(states_);
+  states_.resize(kernels_.size());
+  if (!is_last) {
+    RETURN_NOT_OK(ResetKernelStates());
+  }
```

Comment:
Without this change, any error in `MergeAll` below could lead to a crash.

Reply:
Not sure I understand this - shouldn't error in `MergeAll` be bubbled up and stop execution? (Why does it lead to crash?)

Reply:
I don't know... why a new scalar agg function makes an existing acero node having to change the internal logic?

Is there any test failing w/o this part of change? Maybe I can take a look and then tell if this change is right.

Reply:
Any test where the merge step returns an error could be affected. I guess other hash-aggregate functions never fail in their merge step, so this was not exercised before.

Reply:
That makes sense. Any existing one returning error in merge? Or do I have to do some code modification to reproduce the issue?

Reply:
You can look at `PivotDuplicateValues`. Especially the second case in that test ("Duplicate values in different chunks"). That should hopefully do the trick.

Reply:
I tried, no luck, sorry. And from the code, the case `PivotDuplicateValues` is using group by thus shouldn't be executing this code path. I can't speak for the validity of this change w/o truly understanding what the problem is.

I do have a suspect though: this might be race related. And if, a big if, this is the case, I'm afraid your change doesn't fix the race. However there are still many things unclear so please don't take this seriously until I have new findings.

Do you still hit the problem if you revert this part of the change? Or shall we try reverting it (maybe temporarily) to see if CI is good?

Reply:
> And from the code, the case PivotDuplicateValues is using group by thus shouldn't be executing this code path.

Update: I was wrong about this part - the test is indeed doing the scalar agg.

(But unfortunately this doesn't change rest of the my comment.)

Reply:
Hmm, I can't seem to reproduce locally right now, though I _did_ have to change this to have the tests pass at the time. I'm gonna revert temporarily so that we can check CI.

Reply:
> I do have a suspect though: this might be race related. And if, a big if, this is the case, I'm afraid your change doesn't fix the race. However there are still many things unclear so please don't take this seriously until I have new findings.

Update: Turns out my suspect of race won't happen actually.

So I'm at my wits end by now.

Reply:
Okay, even with Valgrind I can't reproduce anymore.